feat(core): add support for two-way bindings on dynamically-created components

crisbeto opened the pull request:
Builds on the changes from #60137 to add support for two-way bindings on dynamically-created components. Example usage:

```typescript
import {createComponent, signal, twoWayBinding} from '@angular/core';

const value = signal('');

createComponent(MyCheckbox, {
  bindings: [
    twoWayBinding('value', value),
  ],
});
```

In the example above the value of `MyCheckbox` and the `value` signal will be kept in sync.

crisbeto commented on packages/core/src/render3/dynamic_bindings.ts:
```diff
@@ -174,3 +176,47 @@ export function outputBinding<T>(eventName: string, listener: (event: T) => unkn
+      kind: 'twoWay',
+      requiredVars: input[BINDING].requiredVars + output[BINDING].requiredVars,
+    },
+    set target(target: unknown) {
```
I was debating whether to make this a setter or change `target` to be a `setTarget` function. I went with the latter, because the only case where it being a function would be useful is two-way bindings.

eneajaho commented on packages/core/test/acceptance/create_component_spec.ts:
This would work on model inputs too I suppose, even though I didn't see any tests for that, right?

crisbeto committed:
feat(core): add support for two-way bindings on dynamically-created components

Builds on the changes from #60137 to add support for two-way bindings on dynamically-created components. Example usage:

```typescript
import {createComponent, signal, twoWayBinding} from '@angular/core';

const value = signal('');

createComponent(MyCheckbox, {
  bindings: [
    twoWayBinding('value', value),
  ],
});
```

In the example above the value of `MyCheckbox` and the `value` signal will be kept in sync.

pkozlowski-opensource commented on packages/core/test/acceptance/create_component_spec.ts:
```diff
@@ -1369,6 +1370,592 @@ describe('createComponent', () => {
+      expect(dirInstance.value).toBe('dir changed');
+    });
+
+    it('should not bind root component two-way bindings to directives', () => {
```
This was a good test to add 👍

pkozlowski-opensource reviewed (approved):
LGTM

Reviewed-for: public-api 
Reviewed-for: fw-general

devversion commented on packages/core/src/render3/dynamic_bindings.ts:
```diff
@@ -174,3 +176,47 @@ export function outputBinding<T>(eventName: string, listener: (event: T) => unkn
+ * @param value Writable signal from which to get the current value and to which to write new
+ * values.
+ *
+ * ### Usage example
```
nit: is it potentially worth to `{@link ...}` to the API that this function is supposed to be used with?

devversion commented on packages/core/src/render3/dynamic_bindings.ts:
```diff
@@ -174,3 +176,47 @@ export function outputBinding<T>(eventName: string, listener: (event: T) => unkn
+ *
+ * ### Usage example
+ * In this example we create an instance of the `MyCheckbox` component and bind to its `value`
+ * input using a two-way binding.
```
Maybe not worth doing, but have we thought about some recommendations for type safety here? e.g. potentially users could do for public model fields

```ts
twoWayBinding('bla', mySignal as TargetComponent['modelField'])
```

devversion reviewed (commented):
LGTM for public API; I know per design doc we've decided that we can't reliably support any type safety here

Reviewed-for: public-api

devversion reviewed (approved):
LGTM for public API; I know per design doc we've decided that we can't reliably support any type safety here

Reviewed-for: public-api

pkozlowski-opensource commented:
This PR was merged into the repository by commit b154fb391142db6139cddad820a9a72838f0f16c.

The changes were merged into the following branches: main
